core/drivers/clk/clk.c

```diff
@@ -0,0 +1,263 @@
+#include <kernel/boot.h>
+#include <kernel/panic.h>
+#include <libfdt.h>
+#include <mm/core_memprot.h>
+#include <platform_config.h>
+#include <stddef.h>
+
+struct clk *clk_alloc(const char *name, const struct clk_ops *ops,
+		      struct clk **parent_clks, size_t parent_count)
+{
+	struct clk *clk = NULL;
+	unsigned int parent = 0;
+
+	clk = calloc(1, sizeof(*clk) + parent_count * sizeof(clk));
+	if (!clk)
+		return NULL;
+
+	if (parent_count) {
+		clk->parents = (struct clk **)(clk + 1);
+		clk->num_parents = parent_count;
+		for (parent = 0; parent < parent_count; parent++)
+			clk->parents[parent] = parent_clks[parent];
+	}
+
+	clk->name = name;
+	clk->ops = ops;
+	clk->lock = SPINLOCK_UNLOCK;
+	refcount_set(&clk->enabled_count, 0);
+
+	return clk;
+}
+
+void clk_free(struct clk *clk)
+{
+	free(clk);
+}
+
+static bool clk_check(struct clk *clk)
+{
+	if (!clk->ops)
+		return false;
+
+	if (clk->ops->set_parent && !clk->ops->get_parent)
+		return false;
+
+	if (clk->num_parents > 1 && !clk->ops->get_parent)
+		return false;
+
+	return true;
+}
+
+static void clk_compute_rate(struct clk *clk)
+{
+	unsigned long parent_rate = 0;
+
+	clk->rate = 0;
+
+	if (clk->parent)
+		parent_rate = clk->parent->rate;
+
+	if (clk->ops->get_rate)
+		clk->rate = clk->ops->get_rate(clk, parent_rate);
+	else
+		clk->rate = parent_rate;
+}
+
+size_t clk_get_num_parents(struct clk *clk)
+{
+	return clk->num_parents;
+}
+
+struct clk *clk_get_parent_by_index(struct clk *clk, size_t pidx)
+{
+	if (pidx >= clk->num_parents)
+		return NULL;
+
+	return clk->parents[pidx];
+}
+
+static void clk_init_parent(struct clk *clk)
+{
+	size_t pidx = 0;
+
+	switch (clk->num_parents) {
+	case 0:
+		break;
+	case 1:
+		clk->parent = clk->parents[0];
+		break;
+	default:
+		pidx = clk->ops->get_parent(clk);
+		if (pidx >= clk->num_parents)
+			panic("get_parent returned an invalid value");
+
+		clk->parent = clk->parents[pidx];
+		break;
+	}
+}
+
+TEE_Result clk_register(struct clk *clk)
+{
+	if (!clk_check(clk))
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	clk_init_parent(clk);
+	clk_compute_rate(clk);
+
+	DMSG("Registered clock %s, freq %ld", clk->name, clk_get_rate(clk));
+
+	return TEE_SUCCESS;
+}
+
+static TEE_Result clk_enable_no_lock(struct clk *clk)
+{
+	TEE_Result res = TEE_ERROR_GENERIC;
+	struct clk *parent;
+
+	if (refcount_inc(&clk->enabled_count))
+		return TEE_SUCCESS;
+
+	parent = clk_get_parent(clk);
+	if (parent)
+		clk_enable(parent);
+
+	if (clk->ops->enable) {
+		res = clk->ops->enable(clk);
+		if (res)
+			return res;
+	}
+
+	refcount_set(&clk->enabled_count, 1);
+
+	return TEE_SUCCESS;
+}
+
+TEE_Result clk_enable(struct clk *clk)
+{
+	TEE_Result res = TEE_ERROR_GENERIC;
+
+	cpu_spin_lock(&clk->lock);
+	res = clk_enable_no_lock(clk);
+	cpu_spin_unlock(&clk->lock);
+
+	return res;
+}
+
+static void clk_disable_no_lock(struct clk *clk)
+{
+	struct clk *parent = NULL;
+
+	if (!refcount_dec(&clk->enabled_count))
+		return;
+
+	if (clk->ops->disable)
```

Comment:
Actually, refcount_dec will assert if the counter value is 0 so I guess this can't be called on an already disabled clock or it will assert. But maybe I missed something.

Reply:
You're right, nevermind.